kafka-broker/src/main/scala/com/lightbend/lagom/internal/broker/kafka/Consumer.scala

```diff
@@ -0,0 +1,131 @@
+import com.lightbend.lagom.javadsl.api.broker.Subscriber
+
+import akka.Done
+import akka.actor.ActorSystem
+import akka.kafka.{ ConsumerMessage, ConsumerSettings, Subscriptions }
+import akka.kafka.scaladsl.{ Consumer => ReactiveConsumer }
+import akka.stream.{ FlowShape, Materializer }
+import akka.stream.javadsl.{ Flow => JFlow, Source => JSource }
+import akka.stream.scaladsl.{ Flow, GraphDSL, Sink, Unzip, Zip }
+
+/**
+ * A Consumer for consuming messages from Kafka using the akka-stream-kafka API. 
+ */
+class Consumer[Message] private (config: KafkaConfig, topicCall: TopicCall[Message], groupId: Subscriber.GroupId, info: ServiceInfo, system: ActorSystem)(implicit mat: Materializer) extends Subscriber[Message] {
+
+  private val log = LoggerFactory.getLogger(classOf[Consumer[_]])
+
+  @throws(classOf[Subscriber.IllegalGroupId])
+  override def withGroupId(groupId: String): Subscriber[Message] = {
+    val newGroupId = {
+      if (groupId == null) {
+        // An empty group id is not allowed by Kafka (see https://issues.apache.org/jira/browse/KAFKA-2648
+        // and https://github.com/akka/reactive-kafka/issues/155)
+        val defaultGroupId = Consumer.GroupId.default(info)
+        log.debug {
+          "Passed a null groupId, but Kafka requires clients to set one (see KAFKA-2648). " +
+            s"Defaulting $this consumer groupId to $defaultGroupId."
+        }
+        defaultGroupId
+      } else Consumer.GroupId(groupId)
+    }
+
+    if (newGroupId.groupId == groupId) this
+    else new Consumer(config, topicCall, newGroupId, info, system)
+  }
+
+  override def atMostOnceSource: JSource[Message, _] = {
+    ReactiveConsumer.atMostOnceSource(consumerSettings, subscription)
+      .map(_.value)
+      .asJava
+  }
+
+  override def atLeastOnce(flow: JFlow[Message, _, _]): CompletionStage[Done] = {
+    // Creating a Source of pair where the first element is a reactive-kafka committable message, 
+    // and the second it's the actual underlying message. Then, the source of pair is splitted into 
+    // two streams, so that the `flow` passed in argument can be applied to the underlying message.
+    // After having applied the `flow`, the two streams are combined back and the processed message's
+    // offset is committed to Kafka.
+    val pairedCommittableSource = ReactiveConsumer.committableSource(consumerSettings, subscription)
+      .map(committableMessage => (committableMessage, committableMessage.record.value))
+
+    val committOffsetFlow = Flow.fromGraph(GraphDSL.create() { implicit builder =>
+      import GraphDSL.Implicits._
+      val unzip = builder.add(Unzip[ConsumerMessage.CommittableMessage[_, _], Message])
+      val zip = builder.add(Zip[ConsumerMessage.CommittableMessage[_, _], Any])
+      // parallelism set to 1 because offset should be committed in order
+      val committer = Flow[(ConsumerMessage.CommittableMessage[_, _], Any)].mapAsync(parallelism = 1) {
```

Comment:
The commit commands are emitted in order to Kafka, it's collection/waiting of the results that are done in parallel.